### ISSUE TITLE
Improve a11y aspect of the code

### ISSUE THREAD

**Issue by crs1138** [fullstack-hy2020/fullstack-hy2020.github.io#2208]
I'm following up the conversation from Telegram channel. It'd be great to improve the overall a11y standards within the lessons and exercises.

Here are my notes so far.

---

### About finding elements

This chapter seems like a missed opportunity to introduce the students to a11y. Each input should have a corresponding `<label>` element and the tests should taken an advantage of the `getByLabelText()` function. Tests written this way will reinforce the effort to implement code with focus on accessibility.

No, not really. It's just the way it's being taught is not leading to good practice in terms of a11y and in terms of reasons for the test. Let's start with the a11y. In terms of accessibility, it is insufficient to rely on placeholders, they might not be read. Each <input /> should have a corresponding <label> element (this can be visually hidden if the design requires that, but for now let's assume no such design requirement exists). User interacts with such a form made out of various input boxes in a way that, they'll look for a label where to fill in a value. The JSX could be something like the following code:

The test would specify the required input with something like this:
```javascript
const username = screen.getByLabelText(/username/i)
const password = screen.getByLabelText(/password/i)
const loginBtn = screen.getByRole('button', {name: /login/i, exact: false})
```

This coerces the developer to write the React code that has a11y already in mind. Notice that not only there will need to be a <label> element for each of the input boxes, but we are testing for the label text using the regex without paying attention to the letter case or whether it has or hasn't got the colon after the text. The code passing the test will then look for example as follows:

```jsx
return (
    <form onSubmit={...}>
        <div>
            <label htmlFor="username">Username: </label>
            <input type="text" id="username" value={...} onChange={...} placeholder="your-email@example.com" />
        </div>
        <div>
            <label htmlFor="password">Password: </label>
            <input type="password" id="password" value={...} onChange={...} />
        </div>
        <button type="submit">Login</button>
    </form>
  )
```

Testing this way is more future proof, because this way of finding element will still be valid if the designer comes next week and says, I don't want visible labels and btw, change the placeholder to match a username syntax like this.

**Comment by crs1138** [comment]
I have also made a suggestion for edit of the section **[Tests for the Togglable component](https://fullstackopen.com/en/part5/testing_react_apps#tests-for-the-togglable-component)** here on Github. But I can't find it anywhere to include that within this issue… Basically, it's more of the same, just with specific changes of the code used in that section of the lesson.

**Comment by vimode** [comment]
Hey I am so glad you brought this up, I am taking this course and have been struggling to understand why right from the beginning the code samples and examples are not created accessible. I just assumed that its to make the course less daunting to beginners which is not a great way either (if the plan is to introduce accessibility later on).

I am in the same part (part5) and was waiting to see what point does the discussion of a11y comes in as till now it does not feel like most of the HTML markup is semantic. 

I think throughout the course the code samples can be improved upon to be more semantic and adhere to a11y standards, this way it wouldn't need to introduce a11y at any point but be something every learned along the way. 

If this is something that is going to be implemented throughout the course, needs help with I would very much like to help and/or provide feedback along the way.

**Comment by adeleke5140** [comment]
@vimode I believe that's a great suggestion. I have some experience in a11y and would also love to contribute.
